ID3v2 BPM tag workaround for values higher than 500

JosepMaJAZ opened the pull request:
Workaround for a case where the BPM tag on ID3v2 has values that represent decimal values.
As the comment in the code says: Some software uses (or used) to write decimated values without comma,
so the number reads as 1352 or 14525 when it is 135.2 or 145.25.

JosepMaJAZ committed:
Workaround for a case where the BPM tag on ID3v2 has values that represent decimal values

ritschwumm commented:
interesting workaround for the problem that BPM values are defined to be integral numbers :)

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -225,8 +225,15 @@ bool parseBpm(TrackMetadata* pTrackMetadata, QString sBpm) {
     if (isBpmValid) {
+        //Some software uses (or used) to write decimated values without comma,
+        //so the number reads as 1352 or 14525 when it is 135.2 or 145.25
+        if (bpmValue > 5000.0) {
```
When re-interpreting tags in a non-standard way we should log a warning or at least a debug message.

uklotzde replied:
Btw, all bpm values will be rounded to integers when writing ID3v2 tags:

https://github.com/mixxxdj/mixxx/pull/728

That's the standard and we have to follow it. Otherwise you need to keep track internally of the determined bpm scale factor for each track.

JosepMaJAZ replied:
Ok, that's fine. I will add the log message.

The problem with this tag is that when Mixxx reads it, the bpm analyzer does not run for such tracks (it thinks it is already calculated and one has to tell Mixxx to reset the value to force the analyzer to run).
As such, i tried, to make the problem less problematic.

And since you mention writing the tags back, i didn't modify that part (so it still writes it in the standard way), but we may need to rethink it. 

If a file has 1354  and we read 135.4 and then we store 135, and somehow the user rescans or reloads the data from the file, then it will become 135 and even if this looks like a small change, the beatgrid goes all out of place and given and we are back to the scenario of requiring manual deletion and reanalyzing.

sblaisot replied:
I don't understand why you compare to 5000 when the PR title reads "higher than 500". Is there a multiplier somewhere I didn't catch ?

uklotzde commented on src/track/bpm.cpp:
```diff
@@ -38,6 +38,7 @@ double Bpm::valueFromString(const QString& str, bool* pValid) {
 
 QString Bpm::valueToString(double value) {
     if (isValidValue(value)) {
+        //TODO: Shouldn't this be formatted in some way, instead of letting it output in scientific notation?
```
You might add a comment that this non-localized(!) format is the default for most tag types. Special cases like rounding to integer for ID3v2 are handled separately.

Formatting for display purposes is handled by Track::getBpmText(). Currently this doesn't honour the user's locale settings, see also QLocale::toString(). But before modifying the current behaviour we need to check how manual inputs are handled.

JosepMaJAZ replied:
I am not sure what you ask me to do.

uklotzde commented:
I would recommend to add some tests for this feature in metadatatest.cpp. The two limits should be defined explicitly as constants and used in those tests.

JosepMaJAZ commented:
Now I am no longer sure about the correct place for this fix.

I added it to trackmetadatataglib.cpp because i wanted the fix to be nearer to the cause of the problem (the tag), but I modified a local method (parseBPM) that is used by all the metadata formats.
Then, looking at the metadatatest.cpp file, it tests using the Bpm classs directly, or uses the readTrackMetadata method from metadatataglib (which, hopefully, does not need a file to be present).

So maybe I should move the fix out of parseBPM() into the readTrackMetadataFromID3v2Tag().

And finally, I also found this other thing:

In file track.cpp, the getBpm method of Track is as follow:
```
double Track::getBpm() const {
    double bpm = mixxx::Bpm::kValueUndefined;
    QMutexLocker lock(&m_qMutex);
    if (m_pBeats) {
        // BPM from beat grid overrides BPM from metadata
        // Reason: The BPM value in the metadata might be imprecise,
        // e.g. ID3v2 only supports integer values!
        double beatsBpm = m_pBeats->getBpm();
        if (mixxx::Bpm::isValidValue(beatsBpm)) {
            bpm = beatsBpm;
        }
    }
    return bpm;
}
```
But then, I loaded one of the songs that had the wrong BPM and the deck was using the bad value from the tag ( so much, that the wave overview was all white because the beatgrid was so compressed that it was not possible to see anything else).

So, in the end, the final fix should really be that the metadata BPM value is a temporary value until an analyzer is run to get a truthful value (which as I've said in another comment, doesn't happen).

uklotzde commented:
I already took care of the rounding issues you mentioned. BPM values from 
the beatgrid take precedence over the (possibly) rounded value from the tag. 
The value from the tag is only used initially for tracks without a valid BPM 
value in Beats.

I cannot confirm that values are overwritten from files tags. Neither when 
loading a track on a deck nor when reloading the tags from the file. The BPM 
value in Beats is preserved until you clear it manually and reload the file 
tags.

On 05.02.2017 13:34, Josep Maria Antolín Segura wrote:
>
> Now I am no longer sure about the correct place for this fix.
>
> I added it to trackmetadatataglib.cpp because i wanted the fix to be 
> nearer to the cause of the problem (the tag), but I modified a local 
> method (parseBPM) that is used by all the metadata formats.
> Then, looking at the metadatatest.cpp file, it tests using the Bpm classs 
> directly, or uses the readTrackMetadata method from metadatataglib (which, 
> hopefully, does not need a file to be present).
>
> So maybe I should move the fix out of parseBPM() into the 
> readTrackMetadataFromID3v2Tag().
>
> And finally, I also found this other thing:
>
> In file track.cpp, the getBpm method of Track is as follow:
>
> |double Track::getBpm() const { double bpm = mixxx::Bpm::kValueUndefined; 
> QMutexLocker lock(&m_qMutex); if (m_pBeats) { // BPM from beat grid 
> overrides BPM from metadata // Reason: The BPM value in the metadata might 
> be imprecise, // e.g. ID3v2 only supports integer values! double beatsBpm 
> = m_pBeats->getBpm(); if (mixxx::Bpm::isValidValue(beatsBpm)) { bpm = 
> beatsBpm; } } return bpm; } |
>
> But then, I loaded one of the songs that had the wrong BPM and the deck 
> was using the bad value from the tag ( so much, that the wave overview was 
> all white because the beatgrid was so compressed that it was not possible 
> to see anything else).
>
> So, in the end, the final fix should really be that the metadata BPM value 
> is a temporary value until an analyzer is run to get a truthful value 
> (which as I've said in another comment, doesn't happen).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1173#issuecomment-277516907>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-arxlefsn8TtuI_XK9elQ5zNU26uJks5rZcHCgaJpZM4L3TB7>.
>

uklotzde commented:
Don't mind. I will try to improve the documentation of this class and its 
methods later.

On 05.02.2017 12:44, Josep Maria Antolín Segura wrote:
>
> *@JosepMaJAZ* commented on this pull request.
>
> ----------------------------------------------------------------------------
>
> In src/track/bpm.cpp <https://github.com/mixxxdj/mixxx/pull/1173>:
>
> > @@ -38,6 +38,7 @@ double Bpm::valueFromString(const QString& str, bool* pValid) {
>   
>   QString Bpm::valueToString(double value) {
>       if (isValidValue(value)) {
> +        //TODO: Shouldn't this be formatted in some way, instead of letting it output in scientific notation?
>
> I am not sure what you ask me to do.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1173>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-ajNOHhk4slruViRyQRPUX03de5wNks5rZbYggaJpZM4L3TB7>.
>

JosepMaJAZ committed:
Moved the check code strictly to the TBPM ID3v2 tag reading and added testcase

JosepMaJAZ committed:
restoring const, now that the code is elsewhere

uklotzde commented:
Missing include: #include "util/memory.h"

JosepMaJAZ committed:
compilation fix

JosepMaJAZ committed:
ok, this time i did test=1

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -1026,6 +1026,19 @@ void readTrackMetadataFromID3v2Tag(TrackMetadata* pTrackMetadata,
         parseBpm(pTrackMetadata, toQStringFirstNotEmpty(bpmFrame));
+        long bpmValue = pTrackMetadata->getBpm().getValue();
+        //Some software uses (or used) to write decimated values without comma,
+        //so the number reads as 1352 or 14525 when it is 135.2 or 145.25
```
could you name these softwares here?

JosepMaJAZ replied:
I believe it might have been when I had BPM Studio 4, around the year 2003.
I have been able to locate the files that have these tags (now that I know that it's a tag) and it's only 113 file and they are mostly of the years 2000 to 2003.

Btw, I found one file (not related to the others, since it's from 2007) that has this value: "130.01" , yes, with one dot.

daschuer commented on src/track/bpm.h:
```diff
@@ -12,6 +12,8 @@ class Bpm final {
     static constexpr double kValueUndefined = 0.0;
     static constexpr double kValueMin = 0.0; // lower bound (exclusive)
+    static constexpr double kValueMinForTwoDecimals = 5000.0;
+    static constexpr double kValueMinForOneDecimal = 500.0;
```
We should put here a good reference why 50 to 500 is a good BPM range. What is the border of the other tools? 
My Metronom has a range of 40 ...  208 Bpm
This is also confirmed in this German Wikipedia article:
https://de.wikipedia.org/wiki/Tempo_(Musik)#Metronomangaben

How about divide the Bpm until it is in the range of 30 .. 300 BPM?

JosepMaJAZ replied:
Ok. sounds good to me.  (200 would be a bit on the edge, I have a few 190BPM tracks).

daschuer commented:
Ah OK, thinking about his again. This PR is around the question what should happen with a 1500 bpm track? 
This value unusable without a doubt, independent from any APP that is writing it. 
If we consider all file BPM values as not trust worthy anyway it is only a soring issue in the library.
Interpreting it as 150.0 is not a bad option.

JosepMaJAZ committed:
Changing range from 50-500 to 30-300

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -1026,6 +1026,19 @@ void readTrackMetadataFromID3v2Tag(TrackMetadata* pTrackMetadata,
+                pTrackMetadata->getTitle() << " from " << bpmValue << " to " << bpmValue/10.0;
+            bpmValue /= 10.0;
+        }
+        pTrackMetadata->getBpm().setValue(bpmValue);
```
If we one day have the write back loop working this might change the ID3 tag probably unwanted. 
I think it is better to keep the original value, and use the tweaked value for use inside Mixxx only. 
@uklotzde: Or is this not an issue here? 

The two branches above can be turned into a while loop. 

```
while (bpmValue > Bpm::kValueMax) {
    bpmValue \= 10.0;  
}
```

JosepMaJAZ replied:
Ok, i'll rewrite it as a loop (i was considering it, but we didn't had a max value defined (in fact, there's one defined in the test case, but not in the bpm object)).

As for the tag... The reason I wanted to change it is because it affected Mixxx usage (from bad ordering in the library to showing an unrecognizable beatgrid). 
So, what I had to do inevitably was to correct the BPM value by hand.
And If I need to change the value, I am at the same case than with the automatic change if ID3 writing comes back.
In this regard, I consider that worrying about writing a different value than the original woudn't matter (More so with uklotzde comment below).

uklotzde commented:
We would need to store the desired scaling factor for each track, if we want 
to preserve those values for writing.

But I'm not sure if it is really worth to taint our database for writing 
non-standard ID3v2 tags targeted at some obscure external applications. If 
you enable tag writing in Mixxx you will get tags conforming to the official 
standards. No exceptions, no surprises ;)

On 08.02.2017 10:31, Daniel Schürmann wrote:
>
> *@daschuer* commented on this pull request.
>
> ----------------------------------------------------------------------------
>
> In src/track/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/1173#pullrequestreview-20714177>:
>
> > @@ -1026,6 +1026,19 @@ void readTrackMetadataFromID3v2Tag(TrackMetadata* pTrackMetadata,
>       const TagLib::ID3v2::FrameList bpmFrame(tag.frameListMap()["TBPM"]);
>       if (!bpmFrame.isEmpty()) {
>           parseBpm(pTrackMetadata, toQStringFirstNotEmpty(bpmFrame));
> +        long bpmValue = pTrackMetadata->getBpm().getValue();
> +        //Some software uses (or used) to write decimated values without comma,
> +        //so the number reads as 1352 or 14525 when it is 135.2 or 145.25
> +        if (bpmValue >= Bpm::kValueMinForTwoDecimals) {
> +            qWarning() << " Changing BPM on " << pTrackMetadata->getArtist() << " - " <<
> +                pTrackMetadata->getTitle() << " from " << bpmValue << " to " << bpmValue/100.0;
> +            bpmValue /= 100.0;
> +        } else if (bpmValue >= Bpm::kValueMinForOneDecimal) {
> +            qWarning() << " Changing BPM on " << pTrackMetadata->getArtist() << " - " <<
> +                pTrackMetadata->getTitle() << " from " << bpmValue << " to " << bpmValue/10.0;
> +            bpmValue /= 10.0;
> +        }
> +        pTrackMetadata->getBpm().setValue(bpmValue);
>
> If we one day have the write back loop working this might change the ID3 
> tag probably unwanted.
> I think it is better to keep the original value, and use the tweaked value 
> for use inside Mixxx only.
> @uklotzde <https://github.com/uklotzde>: Or is this not an issue here?
>
> The two branches above can be turned into a while loop.
>
> |while (bpmValue > Bpm::kValueMax) { bpmValue \= 10.0; } |
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1173#pullrequestreview-20714177>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-apmo5WPUZLBafzvLTxnVxKLOfOgSks5raYtYgaJpZM4L3TB7>.
>

daschuer commented:
> If you enable tag writing in Mixxx you will get tags conforming to the official standards. No exceptions, no surprises ;)

I agree with that. My concerns are more for the situation where the user wants to change the title or so, will it always "correct" the Bpm value as well? 

Sure, the issue is exotic, because the user will have analysis results in almost all cases. 
Could this become a real issue?

JosepMaJAZ committed:
changed limits to a single kvaluemax and using a loop to divide the read value

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -1026,6 +1026,18 @@ void readTrackMetadataFromID3v2Tag(TrackMetadata* pTrackMetadata,
     const TagLib::ID3v2::FrameList bpmFrame(tag.frameListMap()["TBPM"]);
     if (!bpmFrame.isEmpty()) {
         parseBpm(pTrackMetadata, toQStringFirstNotEmpty(bpmFrame));
+        long bpmValue = pTrackMetadata->getBpm().getValue();
```
getValue() returns a double, why is it passed to a long? 
In general we should avoid long, because it means different things on Windows a Linux.

JosepMaJAZ replied:
ouch! when i moved the code here i changed the type. Fixing it now.

JosepMaJAZ committed:
reverting back to double type. oups!

daschuer commented:
A test fails on windows only: 
```
[ RUN      ] MetadataTest.ParseBpmPrecision
 Changing BPM on  ""  -  ""  from  600  to  60 
src\test\metadatatest.cpp(53): error: Value of: expectedValue
  Actual: 60
Expected: trackMetadata.getBpm().getValue()
Which is: 600
 Changing BPM on  ""  -  ""  from  1256  to  125.6 
src\test\metadatatest.cpp(53): error: Value of: expectedValue
  Actual: 125.59999999999999
Expected: trackMetadata.getBpm().getValue()
Which is: 1256
 Changing BPM on  ""  -  ""  from  14525  to  145.25 
src\test\metadatatest.cpp(53): error: Value of: expectedValue
  Actual: 145.25
Expected: trackMetadata.getBpm().getValue()
Which is: 14525
[  FAILED  ] MetadataTest.ParseBpmPrecision (1 ms)
[ RUN      ] MetadataTest.ParseBpmValidRange
[       OK ] MetadataTest.ParseBpmValidRange (63 ms)
```

JosepMaJAZ commented:
Now that is weird.
Looking at the code, the value of "trackMetadata.getBpm().getValue()" has to be the converted one (because we see the message printed that it has been changed, and such value is set to it afterwards).
Said that, Maybe I see what happens. I might have written the test in the wrong order and when that is compiled, it might get a constant value or something:
        EXPECT_DOUBLE_EQ(trackMetadata.getBpm().getValue(), expectedValue);
Going to change that to:
        EXPECT_DOUBLE_EQ(expectedValue, trackMetadata.getBpm().getValue());
http://www.ibm.com/developerworks/aix/library/au-googletestingframework.html#list9

While looking into that issue, I also found the file src\test\rescalertest.cpp that has the test parameters inverted too. Someone can look at that file and fix it too.

JosepMaJAZ committed:
change to correct the test restult

JosepMaJAZ commented:
Looks like the test result is correct (in identifying a failure) after all.
The TrackMetadata.getBpm() method returns a new instance of the Bpm object rather than a reference, so the line getBpm().setValue(newValue) that I added to readId3v2Metadata() was not altering trackmetadata.

I've now changed the method signature to const Bpm& getValue() const, like the rest of the getter methods of that class, so that now it will detect the error on compile time. 

Going to push the change after compilation and test.

JosepMaJAZ committed:
Fix for setBpm

daschuer commented:
LGTM! 

Thank you.
